### DISCUSSION TITLE
How to make the runtime API call error out during the execution when the recorded storage proof size exceeds the maximum proof size

### DISCUSSION THREAD

**Discussion by liuchengxu** [paritytech/substrate#14350] [Q&A]
In Subspace, we employ an optimistic rollup solution for the execution chain. When a dispute arises, the operator generates a fraud proof, which is then sent to the consensus layer as an unsigned extrinsic. As is commonly understood, extrinsic inherently carries a size limit. Therefore, it's imperative that the size of the fraud proof does not exceed this limit, ensuring it fits within the extrinsic.

The most substantial portion of the fraud proof is the storage proof recorded during the challenged state transition (InitializeBlock/ApplyExtirnsic/FinalizeBlock). Given that the exact size of the storage proof cannot be determined until execution, we need to limit storage access during regular block execution. If the storage proof, required to verify the transition, exceeds the maximum proof size limit, the execution fails.

https://github.com/paritytech/substrate/blob/689da495a0c0c0c2466fe90a9ea187ce56760f2d/primitives/trie/src/recorder.rs#L235-L238

If I'm not mistaken, we currently have the ability to determine the exact proof size using the existing `TrieRecorder.encoded_size_estimation`. Our challenge now is figuring out how to trigger an error during runtime API execution when the proof size surpasses its limit. Do you have any insights on the best way to tackle this? Also, we would be more than willing to contribute to Substrate if this feature is considered broadly beneficial.

**Comment by ggwpez** [comment]
>Given that the exact size of the storage proof cannot be determined until execution, we need to limit storage access during regular block execution.

Sure things like `on_initialize` could over use weight, but that can be prevented by benchmarking them.  
For normal extrinsics you can just set the PoV weight of the block to something smaller. Are you using chromatic weights in your runtime?

>Our challenge now is figuring out how to trigger an error during runtime API execution when the proof size surpasses its limit.

I dont understand where the proof is verified. Why is it verified in a runtime API and not an extrinsic?  
So is the problem in the proof creation or its verification?

**Reply by liuchengxu** [reply]
There are a few downsides to weight:

- Predetermined benchmark results are static, failing to accurately capture the dynamic nature of storage reads/writes for each specific transaction. This lack of precision is less than optimal.
- The weight info is implicitly trusted. If a runtime were to supply malicious weight info, the system will be problematic. It's possible in the future in Subspace that anyone can upload a runtime, potentially exploiting the weight system to make the chain unfunctional.

Unlike Polkadot, which verifies the state transition validity at the block level, we optimistically verify the transition on the transaction level, so setting the PoV weight for the block does not work for us.

Simply put, the fraud proof verification is done on the client side, by fetching the runtime code and spinning up a wasm VM to re-run the state transition to calculate the post-state root, with data verifiably provided by the proof. If the post-state root mismatches the one declared on chain, then the proof is valid, the result on chain is invalid and subsequently pruned.

To illustrate, to verify the state transition of a transaction, we essentially invoke the `apply_extrinsic(uxt)` runtime API on the verifier side with all the necessary data (pre-state root, `uxt`, proof of the state accessed in this transaction's execution) to exactly re-execute this transaction and get the state root after the execution. We then compare the post-state root to see whether the proof is valid or not. 

The fraud proof is submitted in an extrinsic, so the fraud proof size can not be unbounded, we have to terminate the transaction execution when some transaction's execution exceeds the maximum proof size, otherwise, if a bad operator computes this transaction incorrectly, the honest operator can not report it by sending a fraud proof to the settlement layer in an extrinsic.

There is a lot of background here. All in all, the requirement is to fail the transaction when the size of storage proof for the state accessed within the transaction execution surpasses a predetermined threshold.

**Reply by bkchr** [reply]
> otherwise, if a bad operator computes this transaction incorrectly,

They can still do this as they could just disable the error on running over the proof size (which does not exist). 

In general, it sounds like you are coming here up with design questions of your system ;) It could also happen that a tx is executing an infinite loop. You also need to protect against this. You can not run untrusted code on untrusted machines and then be able to ensure that no bad actor is playing your fraud proof system.

**Reply by liuchengxu** [reply]
>They can still do this as they could just disable the error on running over the proof size (which does not exist).

Yes, a bad operator can tamper with the transaction in any invalid way, but that will produce a wrong state root for sure, which will be challenged by the honest operator.

In essential, this problem exists in the both validity proof and fraud proof system in a different form, Polkadot has the max pov size limit, for optimistic rollup on Ethereum, they must ensure the fraud proof can fit into the Ethereum block and the gas consumed is within the Ethereum's gas limit. In our case, the fraud proof is submitted to the consensus chain in an extrinsic. It's not specific to our system. 

Technically, it's similar to GasMeter for smart contracts, here we want a StorageMeter for runtime that will terminate the execution at some point when a tx attempts to do excessive storage reads/writes. The rough idea is to check the storage proof size after every storage access within the transaction execution, and error out to fail the transaction as regular tx failure if the proof size exceeds the limit:

```
let res = access_storage_with_recorder();

if recorder.estimated_storage_proof_size() > MAX_PROOF_SIZE {
    Err("Out of storage")
} else {
    res
}
```

This looks like a custom TrieBackendWithRecorder, any advice is appreciated.

**Reply by bkchr** [reply]
But who should run this code? Who should error out if the fraud proof would  be to big?

**Reply by liuchengxu** [reply]
- Every operator (_L2 validator_) executes the transactions using the custom TrieBackendWithRecorder that would fail the transaction in such a case the storage proof to verify this transaction execution is huge, a selected operator will submit the execution result to the consensus chain.
- If a bad operator was selected and submitted an incorrect execution result to the consensus chain, the other honest operator will observe this and submit a fraud proof in an extrinsic to the consensus chain accordingly.
- The consensus chain node will receive the fraud proof extrinsic and verify the proof validity by re-executing the fraudulent transaction and comparing the state root after execution. If the proof is valid, pack the fraud proof extrinsic into the next consensus block, prune the bad execution result and punish the bad operator.

So the execution chain nodes always run the code and error out in case of the huge fraud proof, the consensus chain node runs the code only when the fraud proof is submitted.

**Reply by bkchr** [reply]
> The consensus chain node will receive the fraud proof extrinsic

But what if the bad actor included some transaction for that the honest operator can not build a fraud proof that is small enough to fit into the extrinsic?

**Reply by nazar-pc** [reply]
> But what if the bad actor included some transaction for that the honest operator can not build a fraud proof that is small enough to fit into the extrinsic?

This is exactly the point here: we need to detect this and abort execution. Then fraud proof can include amount of data that is right above configured read limit and anyone would be able to verify that transaction does not terminate before that, while clearly already exceeding allowed limit.

**Reply by bkchr** [reply]
Okay, that sounds already a little bit different. You would then check that the consensus node has read all the data of the fraud proof, but the transaction would still try to read data that isn't available in the proof?

In general we don't have anything that would error on reading. Probably the best would be to write some `Externalities` wrapper or something similar. Or a modified `TrieBackend`. However, it will be complicated to put the stuff in there on upstream. Will require some tinkering.

**Reply by liuchengxu** [reply]
Yes, but it's actually straightforward for the consensus node as it re-executes the transaction in the same way as the operator does, with the data provided in the fraud proof. Ideally, after reading some storage, the execution should smoothly return an error like OutOfStorage from `apply_extrinsic` API, indicating the tx failure and producing a new state root.

Will have an experiment on our side and then see whether it's possible to upstream back.
